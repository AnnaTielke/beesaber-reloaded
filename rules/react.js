module.exports = {
  plugins: ['react'],

  extends: ['plugin:react/recommended'],

  settings: {
    react: {
      version: 'detect',
    },
  },

  rules: {
    'react/button-has-type': 'error',
    'react/jsx-boolean-value': 'error',
    'react/jsx-fragments': 'error',
    'react/jsx-handler-names': 'error',
    'react/jsx-no-bind': ['error', { allowArrowFunctions: true }],
    'react/jsx-pascal-case': 'error',
    'react/no-access-state-in-setstate': 'error',
    'react/no-array-index-key': 'error',
    'react/no-danger': 'error',
    'react/no-did-mount-set-state': 'error',
    'react/no-did-update-set-state': 'error',
    'react/no-multi-comp': 'error',
    'react/no-this-in-sfc': 'error',
    'react/no-typos': 'error',
    'react/no-unsafe': 'error',
    'react/no-unused-prop-types': 'error',
    'react/no-unused-state': 'error',
    'react/no-will-update-set-state': 'error',
    'react/prefer-es6-class': 'error',
    'react/prefer-stateless-function': 'error',
    'react/style-prop-object': 'error',
    'react/void-dom-elements-no-children': 'error',
  },
};
