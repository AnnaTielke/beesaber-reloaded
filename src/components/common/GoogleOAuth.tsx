import React from 'react';
import styled from 'styled-components';

import { mediaQuery } from '../../tokens/definitions/layout';

const Container = styled.div`
  ${mediaQuery.mobile} {
    margin-top: 20px;
  }
`;

const GoogleOAuth = (): JSX.Element => <Container>Google</Container>;

export default GoogleOAuth;
