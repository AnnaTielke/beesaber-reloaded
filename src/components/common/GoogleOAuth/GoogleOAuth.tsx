/* eslint-disable no-console */
import React from 'react';
import { GoogleLogin, GoogleLogout } from 'react-google-login';
import styled from 'styled-components';

import { errHandler } from '@/api/api';
import { useAppDispatch, useAppSelector } from '@/store/hooks';
import {
  selectIsLoggedIn,
  userIsLoggedIn,
  userIsLogginIn,
} from '@/store/reducer/appStatusReducer';
import { userDataUpdated } from '@/store/reducer/userDataReducer';
import { mediaQuery } from '@/tokens/definitions/layout';

import {
  handleLogin,
  handleLogout,
  loginProps,
  logoutProps,
} from './authHandlers';

import type { PossibleResponses } from '@/api/api';

const Container = styled.div`
  ${mediaQuery.mobile} {
    margin-top: 20px;
  }
`;

const GoogleOAuth = (): JSX.Element | null => {
  const clientId = process.env.REACT_APP_GOOGLE_CLIENTID;

  if (clientId === undefined) return null;
  const dispatch = useAppDispatch();
  const isLoggedIn = useAppSelector(selectIsLoggedIn);

  const onSuccess = async (response: PossibleResponses) => {
    dispatch(userIsLogginIn(true));
    try {
      if ('googleId' in response) {
        const userData = await handleLogin(response);

        dispatch(userDataUpdated(userData));
        dispatch(userIsLoggedIn(true));
      }
    } catch (error: unknown) {
      if (error instanceof Error) {
        errHandler(error);
      }
    }
    dispatch(userIsLogginIn(false));
  };

  const logout = async () => {
    try {
      await handleLogout();
      dispatch(userIsLoggedIn(false));
      dispatch(userDataUpdated({}));
    } catch (error: unknown) {
      if (error instanceof Error) {
        errHandler(error);
      }
    }
  };

  return (
    <Container>
      {!isLoggedIn && (
        <GoogleLogin
          clientId={clientId}
          onSuccess={onSuccess}
          onFailure={() => errHandler(new Error('Google login failed'))}
          {...loginProps}
        />
      )}
      {isLoggedIn && (
        <GoogleLogout
          clientId={clientId}
          onLogoutSuccess={logout}
          {...logoutProps}
        />
      )}
    </Container>
  );
};

export default GoogleOAuth;
