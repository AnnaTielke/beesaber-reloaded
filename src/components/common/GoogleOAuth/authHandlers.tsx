import { toast } from 'react-toastify';

import api from '../../../api/authApi';
import NeonGoogleButton from './NeonGoogleButton/NeonGoogleButton';

import type { GoogleLoginResponse } from 'react-google-login';
import type { UserData } from '../../../sharedTypes/UserData';

type RenderPropsType = {
  disabled?: boolean | undefined;
  onClick: () => void;
};

const loginProps = {
  render: (renderProps: RenderPropsType): JSX.Element => (
    <NeonGoogleButton text={'Login'} {...renderProps} />
  ),
  isSignedIn: true,
  cookiePolicy: 'single_host_origin',
};

const logoutProps = {
  render: (renderProps: RenderPropsType): JSX.Element => (
    <NeonGoogleButton text={'Logout'} {...renderProps} />
  ),
};

const handleLogin = async (
  response: GoogleLoginResponse
): Promise<UserData> => {
  const { googleId, profileObj } = response;
  const userData = await toast.promise<UserData>(
    api.googleLogin({ googleId, profileObj }),
    {
      pending: 'Loggin you in ...',
      success: 'You successfully logged in ðŸ‘Œ',
      error: 'There has been an issue logging you in ðŸ¤¯',
    }
  );

  return userData;
};

const handleLogout = async (): Promise<void> => {
  await toast.promise(api.logout(), {
    pending: 'Loggin you out ...',
    success: 'You successfully logged out ðŸ‘Œ',
    error: 'There has been an issue logging you out ðŸ¤¯',
  });
};

export type { RenderPropsType };

export { handleLogin, handleLogout, loginProps, logoutProps };
