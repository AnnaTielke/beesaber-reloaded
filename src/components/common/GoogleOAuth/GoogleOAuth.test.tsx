import { fireEvent, render, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import GoogleOAuth from './GoogleOAuth';

import type { AppStatus } from '@/store/reducer/appStatusReducer';
import type { RootState } from '@/store/store';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const appStatus: AppStatus = {
  isLoggedIn: true,
  isLoggingIn: false,
};

const store: RootState = {
  appStatus,
  notifications: [],
  userData: {},
};

process.env.REACT_APP_GOOGLE_CLIENTID = 'fakeGoogleClientId';

describe('GoogleOAuth', () => {
  it.each([true, false])(
    'should match snapshot when isLoggedIn=%s',
    (isLoggedIn) => {
      store.appStatus.isLoggedIn = isLoggedIn;
      const { container } = render(
        <Provider store={mockStore(store)}>
          <GoogleOAuth />
        </Provider>
      );

      expect(container).toMatchSnapshot();
    }
  );


  // FIXME: click event does not seem to trigger history push
  it('should redirect, when logging out', async () => {
    const history = createMemoryHistory();
    const path = '/somepath';

    history.push(path);
    const spy = jest.spyOn(history, 'push')

    store.appStatus.isLoggedIn = true;
    render(
      <Router history={history}>
        <Provider store={mockStore(store)}>
          <GoogleOAuth />
        </Provider>
      </Router>
    );

    const logoutButton = await screen.findByRole('button', { name: 'Logout' });

    fireEvent.click(logoutButton);
    expect(spy).toBeCalledTimes(1)
    expect(history.location.pathname).toBe('/');
  });
});
