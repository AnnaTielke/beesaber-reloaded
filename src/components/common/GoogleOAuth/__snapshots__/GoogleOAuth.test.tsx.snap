// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleOAuth should match snapshot when isLoggedIn=false 1`] = `
.c3 {
  color: #edf5ff;
  -webkit-transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  -moz-transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  -o-transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  -webkit-transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  text-shadow: blue -2px -1px 5px,blue 2px 1px 5px, blue 0px 0px 10px,blue 0px 0px 15px,blue 0px 0px 20px, blue 0px 0px 30px;
  font-family: 'NeonTubes2';
  font-weight: 300;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0 10px;
}

.c1 body {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #00111f;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: 'consolas',sans-serif;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0;
  min-height: 100vh;
  padding: 0;
}

.c2 {
  color: rgb(0 0 255);
  display: inline-block;
  font-size: 20px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  overflow: hidden;
  padding: 5px 30px;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c2:hover {
  background: rgb(0 0 255);
  box-shadow: 0 0 10px rgb(0 0 255),0 0 40px rgb(0 0 255), 0 0 80px rgb(0 0 255);
  color: #fff0f0;
  -webkit-transition-delay: 1s;
  transition-delay: 1s;
}

.c2 span {
  display: block;
  position: absolute;
}

.c2 span:nth-child(1) {
  background: linear-gradient(90deg,transparent,rgb(0 0 255));
  height: 2px;
  left: -100%;
  top: 0;
  width: 100%;
}

.c2 span:nth-child(2) {
  background: linear-gradient(270deg,transparent,rgb(0 0 255));
  bottom: 0;
  height: 2px;
  right: -100%;
  width: 100%;
}

.c2 span:nth-child(3) {
  background: linear-gradient(180deg,transparent,rgb(0 0 255));
  height: 100%;
  right: 0%;
  top: -100%;
  width: 2px;
}

.c2 span:nth-child(4) {
  background: linear-gradient(360deg,transparent,rgb(0 0 255));
  bottom: -100%;
  height: 100%;
  left: 0%;
  width: 2px;
}

.c2:hover span:nth-child(1) {
  left: 100%;
  -webkit-transition: 1s;
  transition: 1s;
}

.c2:hover span:nth-child(2) {
  right: 100%;
  -webkit-transition: 1s;
  transition: 1s;
  -webkit-transition-delay: 0.5s;
  transition-delay: 0.5s;
}

.c2:hover span:nth-child(3) {
  top: 100%;
  -webkit-transition: 1s;
  transition: 1s;
  -webkit-transition-delay: 0.25s;
  transition-delay: 0.25s;
}

.c2:hover span:nth-child(4) {
  bottom: 100%;
  -webkit-transition: 1s;
  transition: 1s;
  -webkit-transition-delay: 0.75s;
  transition-delay: 0.75s;
}

@media (max-width:575px) {
  .c0 {
    margin-top: 20px;
  }
}

<div>
  <div
    class="c0"
  >
    <span
      role="button"
    >
      <div
        class="c1"
      >
        <div
          class="c2"
        >
          <span />
          <span />
          <span />
          <span />
          <i
            class="fab fa-google"
          />
           
          <h1
            class="c3"
          >
            Login
          </h1>
        </div>
      </div>
    </span>
  </div>
</div>
`;

exports[`GoogleOAuth should match snapshot when isLoggedIn=true 1`] = `
.c3 {
  color: #edf5ff;
  -webkit-transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  -moz-transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  -o-transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  -webkit-transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  transition: color 0.6s ease-out,text-shadow 0.6s ease-out;
  text-shadow: blue -2px -1px 5px,blue 2px 1px 5px, blue 0px 0px 10px,blue 0px 0px 15px,blue 0px 0px 20px, blue 0px 0px 30px;
  font-family: 'NeonTubes2';
  font-weight: 300;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0 10px;
}

.c1 body {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #00111f;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: 'consolas',sans-serif;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0;
  min-height: 100vh;
  padding: 0;
}

.c2 {
  color: rgb(0 0 255);
  display: inline-block;
  font-size: 20px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  overflow: hidden;
  padding: 5px 30px;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c2:hover {
  background: rgb(0 0 255);
  box-shadow: 0 0 10px rgb(0 0 255),0 0 40px rgb(0 0 255), 0 0 80px rgb(0 0 255);
  color: #fff0f0;
  -webkit-transition-delay: 1s;
  transition-delay: 1s;
}

.c2 span {
  display: block;
  position: absolute;
}

.c2 span:nth-child(1) {
  background: linear-gradient(90deg,transparent,rgb(0 0 255));
  height: 2px;
  left: -100%;
  top: 0;
  width: 100%;
}

.c2 span:nth-child(2) {
  background: linear-gradient(270deg,transparent,rgb(0 0 255));
  bottom: 0;
  height: 2px;
  right: -100%;
  width: 100%;
}

.c2 span:nth-child(3) {
  background: linear-gradient(180deg,transparent,rgb(0 0 255));
  height: 100%;
  right: 0%;
  top: -100%;
  width: 2px;
}

.c2 span:nth-child(4) {
  background: linear-gradient(360deg,transparent,rgb(0 0 255));
  bottom: -100%;
  height: 100%;
  left: 0%;
  width: 2px;
}

.c2:hover span:nth-child(1) {
  left: 100%;
  -webkit-transition: 1s;
  transition: 1s;
}

.c2:hover span:nth-child(2) {
  right: 100%;
  -webkit-transition: 1s;
  transition: 1s;
  -webkit-transition-delay: 0.5s;
  transition-delay: 0.5s;
}

.c2:hover span:nth-child(3) {
  top: 100%;
  -webkit-transition: 1s;
  transition: 1s;
  -webkit-transition-delay: 0.25s;
  transition-delay: 0.25s;
}

.c2:hover span:nth-child(4) {
  bottom: 100%;
  -webkit-transition: 1s;
  transition: 1s;
  -webkit-transition-delay: 0.75s;
  transition-delay: 0.75s;
}

@media (max-width:575px) {
  .c0 {
    margin-top: 20px;
  }
}

<div>
  <div
    class="c0"
  >
    <span
      role="button"
    >
      <div
        class="c1"
      >
        <div
          class="c2"
        >
          <span />
          <span />
          <span />
          <span />
          <i
            class="fab fa-google"
          />
           
          <h1
            class="c3"
          >
            Logout
          </h1>
        </div>
      </div>
    </span>
  </div>
</div>
`;
